activation/activation_test.go

```diff
@@ -1487,3 +1485,99 @@ func TestFindFullyValidHighTickAtx(t *testing.T) {
 		require.Equal(t, atxLower.ID(), found)
 	})
 }
+
+// Test_Builder_RegenerateInitialPost tests the coverage for the edge case
+// where a node operator may change SUs after creating the initial PoST but before
+// submitting the first ATX, which should result in the initial PoST to be deleted
+// and for the new PoST to be generated instead (this also loses the eligibility
+// for the current epoch). This behavior is mocked by mocking the response of the
+// proof validator.
+func Test_Builder_RegenerateInitialPost(t *testing.T) {
+	var (
+		tab           = newTestBuilder(t, 1)
+		sig           = maps.Values(tab.signers)[0]
+		genesis       = time.Now()
+		ctx, cancel   = context.WithCancel(context.Background())
+		commitmentATX = types.RandomATXID()
+		nonce         = types.VRFPostIndex(rand.Uint64())
+		numUnits      = uint32(12)
+		initialPost   = &types.Post{
+			Nonce:   rand.Uint32(),
+			Indices: types.RandomBytes(10),
+			Pow:     rand.Uint64(),
+		}
+	)
+
+	postV2Calls := 0
+	tab.mValidator.EXPECT().
+		PostV2(gomock.Any(), sig.NodeID(), commitmentATX, initialPost, shared.ZeroChallenge, numUnits).DoAndReturn(func(_ context.Context, _ types.NodeID, _ types.ATXID, _ *types.Post, _ []byte, _ uint32, _ ...validatorOption) error {
+		switch postV2Calls {
+		case 0, 1, 3:
+			postV2Calls++
+			return nil
+		case 2:
+			postV2Calls++
+			return nil
+		}
+		panic("unexpected test branch")
+	}).Times(4)
+
+	tab.mclock.EXPECT().CurrentLayer().Return(types.LayerID(0)).AnyTimes()
+	tab.mclock.EXPECT().AwaitLayer(gomock.Any()).DoAndReturn(func(id types.LayerID) <-chan struct{} {
+		// this is our way to end the test - the code will wait on AwaitLayer in PublishActivationTx while
+		// waiting for the publication epoch. otherwise the `run` method will keep on looping to broadcast
+		// future atxs
+		cancel()
+		return make(chan struct{})
+	}).AnyTimes()
+	tab.mclock.EXPECT().LayerToTime(gomock.Any()).DoAndReturn(func(lid types.LayerID) time.Time {
+		// layer duration is 10ms to speed up test
+		return genesis.Add(time.Duration(lid) * 20 * time.Millisecond)
+	}).AnyTimes()
+	tab.mnipost.EXPECT().ResetState(sig.NodeID()).Return(nil).Times(1)
+
+	nipostCalls := 0
+	tab.mnipost.EXPECT().
+		BuildNIPost(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+		DoAndReturn(func(context.Context, *signing.EdSigner, types.EpochID, types.Hash32, *types.NIPostChallenge) (*nipost.NIPostState, error) {
+			switch nipostCalls {
+			case 0:
+				nipostCalls++
+				return nil, ErrInvalidInitialPost
+			case 1:
+				nipostCalls++
+				return &nipost.NIPostState{}, nil
+			default:
+				panic("unexpected branch")
+			}
+
+		}).AnyTimes()
```

Comment:
Instead of manually tracking the calls to the mock you can do the following:

```suggestion
	tab.mnipost.EXPECT().
		BuildNIPost(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
		Return(nil, ErrInvalidInitialPost)
	tab.mnipost.EXPECT().
		BuildNIPost(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
		Return(&nipost.NIPostState{}, nil)
```

This will set up the mock that the first call will return an error, while the second will return `&nipost.NIPostState{}`. Any additional call will cause the test to fail with a message saying that the method of the mock was called unexpectedly.